src/libraries/System.Net.Http/src/System/Net/Http/Metrics/HttpMetricsEnrichmentContext.cs

```diff
@@ -0,0 +1,162 @@
+        {
+            return;
+        }
+
+        _request = request;
+        _response = response;
+        _exception = exception;
+        _tags._tags.Clear();
+        _tags._runningCallback = true;
+
+        try
+        {
+            foreach (Action<HttpMetricsEnrichmentContext> callback in callbackCollection!)
+            {
+                callback(this);
+            }
+        }
+        finally
+        {
+            _tags._runningCallback = false;
+            _request = null;
+            _response = null;
+            _exception = null;
+        }
+        ref TagList enrichmentTags = ref _tags._tags;
+        if (enrichmentTags.Count > 0)
+        {
+            // TagList enumerator allocates, see https://github.com/dotnet/runtime/issues/87022, iterating using an indexer.
+            for (int i = 0; i < enrichmentTags.Count; i++)
+            {
+                tags.Add(enrichmentTags[i]);
+            }
+        }
+    }
+
+    private void EnsureRunningCallback()
+    {
+        if (_request == null) throw new InvalidOperationException("Enrichment callback should not cache HttpMetricsEnrichmentContext");
+    }
+
+    private sealed class TagCollection : ICollection<KeyValuePair<string, object?>>
+    {
+        public bool _runningCallback;
+        internal TagList _tags;
+
+        public int Count
+        {
+            get
+            {
+                EnsureRunningCallback();
+                return _tags.Count;
+            }
+        }
+
+        public bool IsReadOnly
+        {
+            get
+            {
+                EnsureRunningCallback();
+                return _tags.IsReadOnly;
+            }
+        }
+
+        public void Add(KeyValuePair<string, object?> item)
+        {
+            EnsureRunningCallback();
+            _tags.Add(item);
+        }
+
+        public void Clear()
+        {
+            EnsureRunningCallback();
+            _tags.Clear();
+        }
+
+        public bool Contains(KeyValuePair<string, object?> item)
+        {
+            EnsureRunningCallback();
+            return _tags.Contains(item);
+        }
+
+        public void CopyTo(KeyValuePair<string, object?>[] array, int arrayIndex)
+        {
+            EnsureRunningCallback();
+            _tags.CopyTo(array, arrayIndex);
+        }
+
+        public IEnumerator<KeyValuePair<string, object?>> GetEnumerator()
+        {
+            EnsureRunningCallback();
+            return _tags.GetEnumerator();
+        }
+
+        public bool Remove(KeyValuePair<string, object?> item)
+        {
+            EnsureRunningCallback();
+            return _tags.Remove(item);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        private void EnsureRunningCallback()
+        {
+            if (!_runningCallback) throw new InvalidOperationException("Enrichment callback should not cache HttpMetricsEnrichmentContext");
+        }
+    }
```

Comment:
Why is this needed?  If the goal is to ensure that the callback doesn't stash the instance away somewhere, a) it doesn't fully protect against that, and b) we don't try to protect against such things elsewhere (e.g. `List<T>` will pass its internal array to an `ICollection<T>` as part of AddRange).

Reply:
The user can potentially stash a potentially our cached context or collection, and alter it outside of the callback, which may corrupt global behavior since the cache context is shared per-thread. This is a best-effort attempt to prevent that by making such stashed collections useless in most cases.

If we think this is an unlikely misuse, and such protection is unnecessary, I'm happy to simplify.

Reply:
> The user can potentially stash our cached context and/or its' collection

Yes, which is why I drew a parallel to `List<T>`, which hands out its internal array as part of an AddRange call; bad things could happen if the source `ICollection<T>` were to stash that array.  But it's in the camp of "don't do that".  Same as returning an array to the ArrayPool and then continuing to party on it. "Don't do that."